pkg/etcd-backup-operator/controller/s3backup.go

```diff
@@ -0,0 +1,233 @@
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	S3V1                 = "v1"
+	operatorRoot         = "/var/tmp/etcd-backup-operator"
+	backupFilenameSuffix = "etcd.backup"
+)
+
+func (b *Backup) handleS3(clusterName string, s3 *api.S3Source) error {
+	s3b := &s3Backup{
+		namespace:   b.namespace,
+		name:        b.name,
+		clusterName: clusterName,
+		s3bucket:    s3.S3Bucket,
+		prefix:      ToS3Prefix(s3.Prefix, b.namespace, clusterName),
+		awsSecret:   s3.AWSSecret,
+		kubecli:     b.kubecli,
+	}
+	return s3b.saveSnap()
+}
+
+// ToS3Prefix concatenates s3Prefix, S3V1, namespace, clusterName to a single s3 prefix.
+// the concatenated prefix determines the location of S3 backup files.
+func ToS3Prefix(s3Prefix, namespace, clusterName string) string {
+	return path.Join(s3Prefix, S3V1, namespace, clusterName)
+}
+
+type s3Backup struct {
+	namespace   string
+	name        string
+	clusterName string
+
+	s3bucket  string
+	prefix    string
+	awsSecret string
+
+	kubecli kubernetes.Interface
+}
+
+func (b *s3Backup) saveSnap() error {
+	podList, err := b.kubecli.Core().Pods(b.namespace).List(k8sutil.ClusterListOpt(b.clusterName))
+	if err != nil {
+		return err
+	}
+	var pods []*v1.Pod
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		if pod.Status.Phase == v1.PodRunning {
+			pods = append(pods, pod)
+		}
+	}
+
+	if len(pods) == 0 {
+		msg := "no running etcd pods found"
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+	member, rev := getMemberWithMaxRev(pods, nil)
+	if member == nil {
+		logrus.Warning("no reachable member")
+		return fmt.Errorf("no reachable member")
+	}
+
+	log.Printf("saving backup for cluster (%s)", b.clusterName)
+	if err := b.writeSnap(member, rev); err != nil {
+		err = fmt.Errorf("write snapshot failed: %v", err)
+		return err
+	}
+	return nil
+}
+
+func (b *s3Backup) writeSnap(m *etcdutil.Member, rev int64) error {
+	cfg := clientv3.Config{
+		Endpoints:   []string{m.ClientURL()},
+		DialTimeout: constants.DefaultDialTimeout,
+		TLS:         nil,
+	}
+	etcdcli, err := clientv3.New(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create etcd client (%v)", err)
+	}
+	defer etcdcli.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	resp, err := etcdcli.Maintenance.Status(ctx, m.ClientURL())
+	cancel()
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	defer cancel()
+	rc, err := etcdcli.Maintenance.Snapshot(ctx)
+	defer rc.Close()
+	if err != nil {
+		return fmt.Errorf("failed to receive snapshot (%v)", err)
+	}
+	return b.saveToS3(resp.Version, rev, rc)
+}
+
+// NewS3Uploader returns a Uploader that is configured with proper aws config and credentials derived from Kubernetes secret.
+func NewS3Uploader(bucket, prefix string, secret *v1.Secret) (*s3manager.Uploader, error) {
```

Comment:
This function does not need to be exported right? So it should be lowercase `newS3Uploader`.

Reply:
yeah, you are right.